src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts

```diff
@@ -361,35 +362,70 @@ export class CommandDetectionCapability extends Disposable implements ICommandDe
 		this._logService.debug('CommandDetectionCapability#handleCommandStart', this._currentCommand.commandStartX, this._currentCommand.commandStartMarker?.line);
 	}
 
-	private _handleCommandStartWindows(): void {
+	private async _handleCommandStartWindows(): Promise<void> {
 		this._currentCommand.commandStartX = this._terminal.buffer.active.cursorX;
 
 		// On Windows track all cursor movements after the command start sequence
 		this._commandMarkers.length = 0;
-		// HACK: Fire command started on the following frame on Windows to allow the cursor
-		// position to update as conpty often prints the sequence on a different line to the
-		// actual line the command started on.
-		timeout(0).then(() => {
-			if (!this._currentCommand.commandExecutedMarker) {
-				this._onCursorMoveListener = this._terminal.onCursorMove(() => {
-					if (this._commandMarkers.length === 0 || this._commandMarkers[this._commandMarkers.length - 1].line !== this._terminal.buffer.active.cursorY) {
-						const marker = this._terminal.registerMarker(0);
-						if (marker) {
-							this._commandMarkers.push(marker);
-						}
-					}
-				});
+
+		const lastCommand = this.commands.at(-1);
+		const terminal = this._terminal;
+
+		function cursorOnCorrectLine(): boolean {
+			if (!lastCommand) {
+				return false;
 			}
-			this._currentCommand.commandStartMarker = this._terminal.registerMarker(0);
-			if (this._currentCommand.commandStartMarker) {
-				const line = this._terminal.buffer.active.getLine(this._currentCommand.commandStartMarker.line);
-				if (line) {
-					this._currentCommand.commandStartLineContent = line.translateToString(true);
+			const cursorYAbsolute = terminal.buffer.active.baseY + terminal.buffer.active.cursorY;
+			// If the cursor position is within the last command, we should poll.
+			const lastCommandYAbsolute = (lastCommand.endMarker ? lastCommand.endMarker.line : lastCommand.marker?.line) ?? -1;
+			return cursorYAbsolute > lastCommandYAbsolute;
+		}
+
+		function lineIsPrompt(): boolean {
+			const line = terminal.buffer.active.getLine(terminal.buffer.active.baseY + terminal.buffer.active.cursorY);
+			if (!line) {
+				return false;
+			}
+			// TODO: fine tune prompt regex to accomodate for unique configurtions.
+			return line.translateToString(true)?.match(/^(PS.+>)|([A-Z]:\\.*>)/) !== null;
+		}
+
+		// Conpty could have the wrong cursor position at this point.
+		if (!cursorOnCorrectLine() || !lineIsPrompt()) {
+			// Poll for 200ms until the cursor position is correct.
+			for (let i = 0; i < 20; i++) {
+				await timeout(10);
+				if (cursorOnCorrectLine() && lineIsPrompt()) {
+					this._logService.debug('CommandDetectionCapability#_handleCommandStartWindows polling attempts required: ', i + 1);
+					break;
 				}
```

Comment:
I think the main interesting thing to log here is when it does not end up breaking after the loop has finished. Move the log after the for loop and check `i >= 20`?